### PR TITLE
Implementing shareKey (partial transfers)

### LINKED ISSUE #5110: Transfer fee to be paid in time
**Is your feature request related to a problem? Please describe.**
ATM we charge a transferFee which is defined in terms of the current keyPrice.  

**Describe the solution you'd like**
We'd like to switch to reduce the total time remaining on transfer instead.

**Describe alternatives you've considered**

**Additional context**
With this we can switch some APIs back to non-payable, maybe even start using the ZOS ERC-721 files going forward.

### PR DESCRIPTION
**Pull request by nfurfaro**
# Description
It's all in the title.

Possible changes:
- [ ] add/switch to `safeShareKey` following example of `safeTransferFrom`

# Issues

Fixes #5110 
Refs #

# Checklist:

- [x] 1 PR, 1 purpose: my Pull Request applies to a single purpose
  - [ ] This PR only contains configuration changes (package.json, etc.)
  - [x] This PR only contains code changes (if configuration changes are required, do a separate PR first, then re-base)
- [x] My code follows the style guidelines of this project, including naming conventions
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] If my code adds or changes components, I have written corresponding stories with Storybook
- [x] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

File: smart-contracts/contracts/mixins/MixinTransfer.sol [hunk 1]
```diff
@@ -44,6 +44,65 @@ contract MixinTransfer is
...
+      address(0), // This is a creation or time-sharing
+      _to,
+      IdTo
+    );
+  }
+
+  /**
+   * This is payable because at some point we want to allow the LOCK to capture a fee on 2ndary
+   * market transactions...
+   */
```
Comment (nfurfaro) [inline smart-contracts/contracts/mixins/MixinTransfer.sol]: oops. not sure why this is back but I will remove it.

File: smart-contracts/contracts/mixins/MixinTransfer.sol [hunk 2]
```diff
@@ -44,6 +44,65 @@ contract MixinTransfer is
...
+      fromKey.expirationTimestamp = block.timestamp; // Effectively expiring the key
+      emit ExpireKey(_tokenId);
+    }
+
+    if (toKey.tokenId == 0) {
+      _assignNewTokenId(toKey);
+      _recordOwner(_to, IdTo);
+    }
+    // add time to new key
+    _timeMachine(IdTo, time, true);
```
Comment (NickCuso) [inline smart-contracts/contracts/mixins/MixinTransfer.sol]: Hmm this is transferring `_timeShared + fee` to the other account?  For this maybe we should add a test that confirms the total time remaining across both accounts is <= the original time remaining - fee

Comment (nfurfaro) [reply smart-contracts/contracts/mixins/MixinTransfer.sol]: Good catch. I've reworked this to handle this better, but I'll add a test case as you suggest as well.

File: smart-contracts/contracts/mixins/MixinTransfer.sol [hunk 3]
```diff
@@ -44,6 +44,62 @@ contract MixinTransfer is
...
+  /**
+  * @notice Allows the key owner to share their key (parent key) by
+  * transferring a portion of the remaining time to a new key (child key).
+  * @param _from The owner of the parent key
+  * @param _to The recipient of the shared key
+  * @param _tokenId the key to share
+  * @param _timeShared The amount of time shared
+  */
+  function shareKey(
+    address _from,
```
Comment (NickCuso) [inline smart-contracts/contracts/mixins/MixinTransfer.sol]: Why accept _from?

By using onlyKeyOwnerOrApproved you are confirming the msg.sender has access to that tokenId, but then in the code below you pull the key indexed by _from.

File: smart-contracts/contracts/mixins/MixinTransfer.sol [hunk 4]
```diff
@@ -44,6 +44,62 @@ contract MixinTransfer is
...
+  ) public
+    onlyIfAlive
+    hasValidKey(_from)
+    onlyKeyOwnerOrApproved(_tokenId)
+  {
+    require(_to != address(0), 'INVALID_ADDRESS');
+
+    Key storage fromKey = keyByOwner[_from];
+    Key storage toKey = keyByOwner[_to];
+    uint IdTo = toKey.tokenId;
```
Comment (NickCuso) [inline smart-contracts/contracts/mixins/MixinTransfer.sol]: minor: style convention, not sure we ever start a variable with a capital letter..
where's ESlint on this?!  heh I still wonder if that's actually running at all sometimes.

File: smart-contracts/contracts/mixins/MixinTransfer.sol [hunk 5]
```diff
@@ -44,6 +44,62 @@ contract MixinTransfer is
...
+    uint time;
+    // get the remaining time for the origin key
+    uint timeRemaining = fromKey.expirationTimestamp - block.timestamp;
+    // get the transfer fee
+    uint fee = getTransferFee(_from, _timeShared);
+
+    // ensure that we don't try to share too much
+    if(_timeShared.add(fee) < timeRemaining) {
+      time = _timeShared;
```
Comment (NickCuso) [inline smart-contracts/contracts/mixins/MixinTransfer.sol]: This results in overcrediting the other account still, right?  If I try to transfer 1 year to a friend but only have 1 day remaining, it seems I'm charged a day but they still got they full year (?)

File: smart-contracts/contracts/mixins/MixinTransfer.sol [hunk 6]
```diff
@@ -44,6 +44,62 @@ contract MixinTransfer is
...
+
+    if (toKey.tokenId == 0) {
+      _assignNewTokenId(toKey);
+      _recordOwner(_to, IdTo);
+    }
+    // add time to new key
+    _timeMachine(IdTo, time, true);
+    // trigger event
+    emit Transfer(
+      address(0), // This is a creation or time-sharing
```
Comment (NickCuso) [inline smart-contracts/contracts/mixins/MixinTransfer.sol]: Hmm is from address(0) best here?  maybe we should use the from account?

File: smart-contracts/contracts/mixins/MixinTransfer.sol (hunk 3 above)
Comment (nfurfaro) [reply smart-contracts/contracts/mixins/MixinTransfer.sol]: Well, first, we want to make sure that `_from` has a valid key before trying to share it, and second, with from we can look up the key in one step, vs using `_tokenId` to look up the owner, then looking up the key.

File: smart-contracts/contracts/mixins/MixinTransfer.sol (hunk 4 above)
Comment (nfurfaro) [reply smart-contracts/contracts/mixins/MixinTransfer.sol]: oops, my bad. fixed.

File: smart-contracts/contracts/mixins/MixinTransfer.sol (hunk 3 above)
Comment (NickCuso) [reply smart-contracts/contracts/mixins/MixinTransfer.sol]: if _from as approved to spend _tokenId (which is owned by someone else), this function proceeds and then _fromKey.tokenId != _tokenId

File: smart-contracts/contracts/mixins/MixinTransfer.sol (hunk 6 above)
Comment (nfurfaro) [reply smart-contracts/contracts/mixins/MixinTransfer.sol]: Yeah, could do. I guess it depends on the recipient. If they don't have a key, a new key is being created for them. If they already have a key, they just get a top-up. 
I really don't like either option...but equally fine with using either.

Comment (NickCuso) [reply smart-contracts/contracts/mixins/MixinTransfer.sol]: I think doing it once or twice makes sense.  If it's a new key, emit from 0 as the spec requires that on creation.  Then always emit from -> to as well at the end.

File: smart-contracts/contracts/mixins/MixinTransfer.sol (hunk 5 above)
Comment (nfurfaro) [reply smart-contracts/contracts/mixins/MixinTransfer.sol]: So, if the time I want to share + fee < the timeRemaining on my key, then `time = _timeShared`, but if the time I want to share + fee >= the timeRemaining on my key, then we expire my key and share the time remaining -fee

Make sense?

Comment (nfurfaro) [reply smart-contracts/contracts/mixins/MixinTransfer.sol]: Basically, the value of `_time` depends on whether I'm trying to share less than I have or more. In your example, "they" only got a day minus the fee.

Comment (NickCuso) [reply smart-contracts/contracts/mixins/MixinTransfer.sol]: yup my bad

File: smart-contracts/contracts/mixins/MixinTransfer.sol (hunk 6 above)
Comment (nfurfaro) [reply smart-contracts/contracts/mixins/MixinTransfer.sol]: That works for me.

File: smart-contracts/contracts/mixins/MixinTransfer.sol (hunk 3 above)
Comment (nfurfaro) [reply smart-contracts/contracts/mixins/MixinTransfer.sol]: I was thinking that even if this function is called by someone else who is approved, `_from` is still the address of the key owner, not msg.sender.

Comment (nfurfaro) [reply smart-contracts/contracts/mixins/MixinTransfer.sol]: We could remove this param, and just use `_tokenId` to look up the owner, and then get the key and check its validity using that...

Comment (nfurfaro) [reply smart-contracts/contracts/mixins/MixinTransfer.sol]: But we can't assume the `_from` address is the message.sender. In fact, as per the interface, `_from` is "The owner of the parent key"

Comment (nfurfaro) [reply smart-contracts/contracts/mixins/MixinTransfer.sol]: We can change this if you like, but it's the exact same pattern we use in `transferFrom`

Comment (nfurfaro) [reply smart-contracts/contracts/mixins/MixinTransfer.sol]: resolved in slack channel.

File: smart-contracts/contracts/interfaces/IPublicLock.sol [hunk 7]
```diff
@@ -450,6 +450,46 @@ interface IPublicLock {
   /// @param _tokenId The NFT to transfer
   function transferFrom(address _from, address _to, uint _tokenId) external;
 
+  /**
+  * @notice Allows the key owner to share their key (parent key) by
+  * transferring a portion of the remaining time to a new key (child key).
+  * @dev Throws if _from does not have a valid key.
+  * @dev Throws if _from does not have enough remaining time on
```
Comment (NickCuso) [inline smart-contracts/contracts/interfaces/IPublicLock.sol]: comments about _from no longer applicable

File: smart-contracts/contracts/mixins/MixinTransfer.sol [hunk 8]
```diff
@@ -44,6 +44,65 @@ contract MixinTransfer is
...
+    uint time;
+    // get the remaining time for the origin key
+    uint timeRemaining = fromKey.expirationTimestamp - block.timestamp;
+    // get the transfer fee
+    uint fee = getTransferFee(keyOwner, _timeShared);
+
+    // ensure that we don't try to share too much
+    if(_timeShared.add(fee) < timeRemaining) {
+      time = _timeShared;
+      uint timePlusFee = _timeShared.add(fee);
```
Comment (NickCuso) [inline smart-contracts/contracts/mixins/MixinTransfer.sol]: Minor: but we already safeMath added these values.  For gas, we could cache the result (or use + here)

File: smart-contracts/contracts/mixins/MixinTransfer.sol [hunk 9]
```diff
@@ -44,6 +44,65 @@ contract MixinTransfer is
...
+      // deduct time from parent key, including transfer fee
+      _timeMachine(_tokenId, timePlusFee, false);
+    } else {
+      fee = getTransferFee(keyOwner, timeRemaining);
```
Comment (NickCuso) [inline smart-contracts/contracts/mixins/MixinTransfer.sol]: Hmm this seems okay but looking at it again, maybe we could simplify by using a standard transfer here.  At this point it should be on-par with the logic there.  We could maybe do that later..

File: smart-contracts/contracts/mixins/MixinTransfer.sol [hunk 10]
```diff
@@ -44,6 +44,65 @@ contract MixinTransfer is
...
+      time = timeRemaining.sub(fee);
```
Comment (NickCuso) [inline smart-contracts/contracts/mixins/MixinTransfer.sol]: For gas savings, we could probably remove the .sub here.  We know that the fee is <= 100% making this safe.

**Comment by nfurfaro** [comment]
Thx for the feedback. I'll address your concerns and merge tonight.

File: smart-contracts/contracts/interfaces/IPublicLock.sol (hunk 7 above)
Comment (nfurfaro) [reply smart-contracts/contracts/interfaces/IPublicLock.sol]: yup. fixed.

File: smart-contracts/contracts/mixins/MixinTransfer.sol (hunk 9 above)
Comment (nfurfaro) [reply smart-contracts/contracts/mixins/MixinTransfer.sol]: Yeah, I had a similar thought about this. 
The one difference here is that shareKey allows the original owner to retain their original tokenId (for an expired key), but maybe that doesn't matter. Maybe something to look at after this release.

File: smart-contracts/contracts/mixins/MixinTransfer.sol (hunk 10 above)
Comment (nfurfaro) [reply smart-contracts/contracts/mixins/MixinTransfer.sol]: yep. done.

File: smart-contracts/contracts/mixins/MixinTransfer.sol (hunk 8 above)
Comment (nfurfaro) [reply smart-contracts/contracts/mixins/MixinTransfer.sol]: done.

**Pull request merged**